### PR TITLE
feat!: Remove python 3.8, upgrade dependencies, and add python 3.12 support

### PR DESCRIPTION
**Pull request by antazoey**
### What I did

fixes: #

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

**Review by fubuloubu — APPROVED** [review]
Looks good, but is there anything driving the 3.8 deprecation?

I don't think we should drop 3.8 before Ape does at the least

**Comment by antazoey** [comment]
@fubuloubu I had thought we wanted to drop 3.8 as part of the 0.8 Ape upgrades.

Eip712 is a dependency of ape and it needs to be at least be 3.12 compatible before Ape. It is just easier to drop 3.8 and add 3.12 at the same time.

File: setup.py [hunk 1]
```diff
@@ -64,13 +65,12 @@
...
-        "eth-hash[pycryptodome]",  # NOTE: Pinned by eth-abi
-        "eth-typing>=3.5.2,<4",
-        "eth-utils>=2.3.1,<3",
-        "hexbytes>=0.3.0,<1",
+        "dataclassy>=0.11.1,<1",
+        "eth-abi>=5.1.0,<6",
+        "eth-account>=0.12.1,<0.13",
+        "eth-typing>=4.1.0,<5",
+        "eth-utils>=4.1.0,<5",
+        "hexbytes>=1.2.0,<2",
```
Comment (antazoey) [inline setup.py:73]: because of this change, thinking of going up a full minor

**Comment by fubuloubu** [comment]
> @fubuloubu I had thought we wanted to drop 3.8 as part of the 0.8 Ape upgrades.
> 
> Eip712 is a dependency of ape and it needs to be at least be 3.12 compatible before Ape. It is just easier to drop 3.8 and add 3.12 at the same time.

I am okay dropping it, just should be minor breaking update so people using ape on 3.8 don't end up with an older package that won't download for them

**Comment by antazoey** [comment]
merged but reverted, and a new pr. i will add 3.12 first like you are suggesting.

**Pull request merged**